src/osd/PG.cc

```diff
@@ -3827,8 +3827,13 @@ void PG::append_log(
      /* We must be a backfill peer, so it's ok if we apply
       * out-of-turn since we won't be considered when
       * determining a min possible last_update.
+      *
+      * We skip_rollforward() here, which advances the crt, without
+      * doing an actual rollforward. This avoids cleaning up entries
+      * from the backend and we do not end up in a situation, where the
+      * object is deleted before we can _merge_object_divergent_entries().
```

Comment:
So if this turns out to be an async-recovery peer, it is possible we'll leave some obsolete gen objects behind (without being able to delete them again)?

Also the comment above needs a revision...

Reply:
We'll rely on scrub to clean them up, hence the change to _scan_rollback_obs(). During check_local() on the primary, we do identify such obsolete objects, but ignore them https://github.com/ceph/ceph/blob/master/src/osd/PrimaryLogPG.cc#L13553-L13554.

Yes, I'll fix the comment to include async_recovery peer too.